Upload attendance feature

jonathanchan2 opened the pull request:
Hi Dan, 

A couple months ago I was tasked with customizing the official moodle attendance plugin for use by my university's department. One of the features they want implemented is an automatic attendance marking via a csv upload. Basically, the same feature a lot of people have been requesting in issue #189 .

Over the past few months, I have been working on developing this feature and I believe I have a workable solution. My solution makes use of a csv file generated by scanning the students' id cards using a barcode scanner. The scanner used by my department is the OPN-2001 and the software used to extract the csv file is "OPN200x". The csv file would contain the student's id number, barcode encoding scheme, scan time, scan date, current time, current date and the scanner's id as shown in the figure below. 

![image](https://user-images.githubusercontent.com/45237038/55892458-5916fc00-5b84-11e9-83dc-6cc3ef087ab2.png)

The csv file would be uploaded under the correct class session and the plugin would automatically update the students' attendance records with the correct attendance status. The user would be warned if they upload: 
- the wrong file/file type
- a file that is affected by daylight savings time 
- a file that is not formatted correctly.

For a more detailed description of the developed feature, you can take a look at my system requirements document which I have attached below. It outlines the need for this feature, how the feature will integrate with current processes and fully details how the feature is expected to work, under subsections 3.1 A, 3.1 B and 3.1 D of the requirements document. Subsection 3.1 A addresses the file to be uploaded. Subsection 3.1 B addresses how the uploaded file is to be processed. Subsection 3.1 D addresses how the attendance records are to be edited.   

[SysRS V3.pdf](https://github.com/danmarsden/moodle-mod_attendance/files/3064972/SysRS.V3.pdf)

Please note that only subsections 3.1 A, 3.1 B and 3.1 D are addressed in the code for this pull request currently. The other subsections under section 3 have not been tackled yet and can be ignored. 

I have tested the plugin with real test data from one of our courses and the plugin has performed as expected, greatly reducing the effort to mark the attendance of classes. My feature was integrated into your most updated plugin version, however, I was only able to test on moodle 3.5 since it was the only version of moodle that was available to me. Below, I have attached a document outlining how to set up and carry out the testing on the feature. 

[Testing Procedure.pdf](https://github.com/danmarsden/moodle-mod_attendance/files/3066385/Testing.Procedure.pdf)

I have also attached a .zip file containing the test data, my testing documentation (Test Report v3.xlsx & FR_TC10.xlsx) and a .csv file (newusers.csv) needed for setting up the test environment. Please note, in the test data folder, files ending with "+" have all been affected by daylight savings time. Some files also contain multiple days of records. If you would like to see the results of my own testing, you can look at pages 1 and 2 of "Test Report v3.xlsx" and "FR_TC10.xlsx".

[Attendance plugin files.zip](https://github.com/danmarsden/moodle-mod_attendance/files/3066383/Attendance.plugin.files.zip)

Thank you in advance for any assistance you are able to provide in getting this feature to be a part of the official attendance plugin. It will be truly appreciated. 

Jonathan Chan.

Contact info:
jonathan.chan@sta.uwi.edu
chan.jonathan1@outlook.com

danmarsden commented:
Hi Johnathan - thanks for sharing - would be really great to see this feature supported.

Couple of initial comments after a really quick flick past the code - it looks like the statues PLEA are hard-coded in there, these statuses are not "fixed" and admins/teachers can set a wide range of status values so we'd need make this more generic.

The travis ci checks show quite a few coding guideline issues with the new code, and also some php errors showing up in the behat tests.

Also - I see some strange "copyright" code comments being added in existing files - these don't meet the normal coding guidelines - typically the "first person" who creates a file or class (not individual function) will add a copyright statement to the file - if there are existing files that have this missing I'd prefer you didn't add them as part of this patch.

Is this something you're able to spend more time tidying up before I take another look?

jonathanchan2 commented:
Hi Dan,

Thank you for your feedback and support. 

I decided to stick with the standard PLEA statuses for the initial release, assuming they would be the most commonly used by lecturers. It also allowed for more control over which statuses would be used to mark students as present, late, excused and absent. However, I understand why these statuses are not intended to be "fixed". How would you suggest I correct this?

This project is my first time developing in moodle and php and I am still learning. So, I am not accustomed to travis ci or behat tests and what would cause them to fail or how to deal with them. 

With respect to the copyright comments, I tried to follow the coding guidelines but I do not completely understand how to show which snippets of code were written by me especially if I made changes to an existing file. If you think my copyright comments violate the coding guidelines, you can remove/adjust them to suit by all means. I am okay with that. Also, if there are things in the code you think can be done better, feel free to let me know. 

I can try to tidy up the code with some of your comments but I also have to continue working on the other parts of my project and splitting my time recently has been difficult. So if you feel like there are things you can change, by all means, go ahead.

danmarsden commented:
Thanks Jonathan! - the commit history of a repository shows who is responsible for the code being added. Copyright headers in phpdocs for Moodle files are only added at the top of the file in the main phpdoc block and then at the top of each class (not at the top of each function)- and the first person who "creates" the file or class will add their details. then all future changes to the file by other people will just be recorded in the git commit history.

I haven't really looked at the code properly yet, but I think you might be hard-coded the status values as a way to check if an update call will change the status and give the user more "points" than they previously had? - (I could be wrong though - I haven't really looked at it properly) 

Typically an "admin" process like an import via csv you would expect that the dataset could be trusted - normally any validation should really happen outside Moodle, but If you want to prevent statuses being changed you could check the grade/point values assigned against each status value being passed and only allow a status value to be set with a lower points value? - alternatively for the first release you could remove that custom code and just assume we can trust the incoming CSV data and then at a later date you could implement some custom handling around this (The site admin should be able to turn this off though - I can see many cases where a CSV update should be allowed to update the status of a session even if the user may be granted a higher grade.)

thanks for working on this!

danmarsden commented:
btw - if you haven't seen it already, you can run a lot of the coding guideline tests locally using this:
https://moodle.org/plugins/local_codechecker

There are also some docs in moodledocs that let you put those rules into popular IDE's like https://docs.moodle.org/dev/Linting#Linting_in_your_editor

danmarsden commented:
the other related plugin which checks phpdoc syntax is here:
https://moodle.org/plugins/local_moodlecheck

danmarsden commented:
also another thing to be aware of - class files like your csv_upload_form need to be inside the "classes" folder and ideally need to be namespaced eg:
namespace mod_attendance\import;

or the class name must be prefixed with the plugin frankenstyle name "mod_attendance" - (bit of inconsistency with the existing code, but good for new stuff to follow moodle guidelines where possible.)

jonathanchan2 commented:
I have fixed most of the travis ci errors with respect to the codechecker and removed the inappropriate copyright handles. I made use of the local codechecker plugin (https://moodle.org/plugins/local_codechecker) and it proved very useful. Thank you for pointing this out. I was also able to move the two class files (csv_upload_form & importer) to the "classes" folder. Took a while to figure out how to do this without breaking the code but I was able to do it and made sure that the code still worked on my moodlesite. It does.  

The travis ci is still showing some errors. One error is the copy/paste detector, however, I need both of those files to execute the duplicated code as they apply to two different scenarios. One scenario is when the attendance form is manually entered the other is when the attendance form is being automatically processed. Is there a way to circumvent this error?

The last set of errors are from the behat test report which I have no experience working with. These files that fail the test are also files I have never touched. I have no idea how to deal with these files/errors. 

With respect to your suggestions on tackling the "fixed" attendance status issue, I will have to get back to you on that after I have more thoroughly thought through the issue and your suggestion.

jonathanchan2 commented:
Hi Dan,

The  reason for the hard-coded status check was so that the code would know which status id to automatically assign to a student if they were "present", "late", "excused" or "absent" based on their scan time. I chose these four statuses because they were the default status set and assumed the user wouldn't have reason to change the names of the default statuses. 

The hard-coded status check was not to see if an update call would change the student's attendance status and prevent that change. The prevention of status changes is implemented as a different feature, to prevent lecturers from changing students records maliciously i.e. changing a student's attendance to absent. However, I do understand that this feature should be allowed to be turned off via the site administrator. I will look into implementing an on/off switch for this feature so that the site administrator can make the executive decision on whether or not to use it. 

Now, with respect to the "fixed" attendance issue, my supervisor and I thought about it and came up with a possible solution. Your opinion on our solution would be greatly appreciated. 

 We decided that we will remove dependency on the status description, i.e. the instances of the hard-coded statuses, by implementing a "rule" on each attendance status in the status set. This rule will determine which attendance status is assigned to the student. There will be three main rules:

1.  "This status applies if the student's scan time occurs either within 30 minutes before the start of class or within the first 15 minutes of class"

2. "This status applies if the student's scan time occurs after the first 15 minutes of class but within the class session"

3. "This status applies if the student's scan time occurs after the end of the class session or does not occur at all"  

By default, rule 1 will apply to students who will be considered as present, rule 2 will apply to students who will be considered as late and rule 3 will apply to students who will be considered as absent. There will be a one to one relationship between each rule and its assigned status, preventing the user from assigning the same rule to multiple statuses with warnings. 

These rules will allow the lecturer/user to rename the default statuses as they see fit while still defining how the statuses will automatically apply to the students when an csv attendance file is uploaded. This will remove the need for the "fixed" attendance statuses/hard-coded implementation. The lecturer/user will be able to define these rules under the "Status Set" tab. 

Thank you again for your continued support.

danmarsden commented:
> The travis ci is still showing some errors. One error is the copy/paste detector, however, I need both of those files to execute the duplicated code as they apply to two different scenarios. One scenario is when the attendance form is manually entered the other is when the attendance form is being automatically processed. Is there a way to circumvent this error?

No sorry - Best way to do this is to create a new function that contains the duplicate code and then call it in both places.

danmarsden commented:
> We decided that we will remove dependency on the status description, i.e. the instances of the hard-coded statuses, by implementing a "rule" on each attendance status in the status set. This rule will determine which attendance status is assigned to the student. There will be three main rules:

I'd prefer you used the existing timing settings in the status set which are already used for student self-marking - I'm not clear why you would need to implement a "new" setting to control this? - Have you seen how the student self-marking currently works with the various status values being available for selection and it chooses the highest graded option that is still available at the time of selection?

I'd probably suggest using the existing core "attendance_session_get_highest_status" function for this - but add the ability for that function to receive the time the attendance was taken as a parameter instead of using the current time() value.

Does that make sense?

danmarsden commented:
one of the issues with the existing "timing" settings in the status set is that it requires student self-marking to be enabled - we need to fix this as part of Issue #365 which shouldn't be too much effort.

jonathanchan2 commented:
Hi Dan, 

I was able to mend the duplicated code issue by adjusting the take_from_form_data() function in the structure file and simply calling it in my upload_attendance file. 

Thank you for giving some feedback on my approach to the "fixed" attendance status issue. I can see where you are coming from. There is no need for a new setting to control the statuses if there are existing functions that can help me solve this problem. Up until your last comment, I did not look into the student self-marking feature as it was not immediately useful in helping me create the automatic attendance upload feature. However, I can see how some of its functions can be useful in helping me iron out this "fixed" status issue. 

You mentioned that there was an issue with the student self-marking feature that needed to be rectified before I could use some of its functions. I looked into issue #365 and made some changes according to your last suggestion in the issue thread. I created a new pull request #384 for this and I hoped that my fixes would be able to fully address issue #365. However, I encountered an error with the travis-ci build which I do not know how to address. Can you please provide some insight on this error?

Best regards, 
Jonathan.

jonathanchan2 commented:
Hi Dan,

Since we last spoke, I was able to create a solution using your suggestion to utilize the "attendance_session_get_highest_status" function and the scan time of each attendance record. I have also tested my new code with the real test data from one of our courses and it performs as expected, according to the functional requirements outlined under subsections 3.1 A, 3.1 B and 3.1 D in my system requirements. At this stage, I have decided to move FR_D2 and FR_D3 in the system requirements to a later phase of the plugin as it is not immediately necessary. So, they may be ignore. 
[SysRS V3.pdf](https://github.com/danmarsden/moodle-mod_attendance/files/3180308/SysRS.V3.pdf)

Please have a look at my code and play around with the test data I have provided and let me know your thoughts. I look forward to getting this feature integrated into the official plugin so that people can begin to make use of the more efficient method of taking attendance.  
[Attendance plugin files 14-5-19.zip](https://github.com/danmarsden/moodle-mod_attendance/files/3180285/Attendance.plugin.files.14-5-19.zip)

Recall, in the test data folder, files ending with "+" have all been affected by daylight savings time. Some files also contain multiple days of records. If you would like to see the results of my own testing, you can look at pages 1 and 2 of "Test Report v3.xlsx" and "FR_TC10_B.xlsx".

I have also attached a document outlining how to set up and carry out the testing on the feature. 
[Testing Procedure 14-5-19.pdf](https://github.com/danmarsden/moodle-mod_attendance/files/3180289/Testing.Procedure.14-5-19.pdf)

N.B. I was not able to figure out the behat test error from the travis-ci tests. I will continue to look into it but I suspect that I may need some help with it.

Thank you for you continued support and assistance with this attendance upload feature. It is greatly appreciated.

Jonathan.

danmarsden commented on renderer.php:
```diff
@@ -826,8 +838,25 @@ private function construct_take_user_controls(attendance_take_data $takedata, $u
             }
 
             $celldata['text'] = array();
+
```
there shouldn't be any hard-coded references to specific statuses like Present/Late/Excused/Absent - haven't looked closely at what this code is doing, but to get this into the core release we can't be hard-coding this stuff.

jonathanchan2 replied:
i agree. this needs to be removed. will work on this.

danmarsden commented on classes/importer.php:
```diff
@@ -0,0 +1,364 @@
+
+        $filename = $CFG->tempdir.'/csvimport/'.'attendance'.'/'.$USER->id.'/'.$this->importid;
+        $tempfilename = $CFG->tempdir.'/csvimport/'.'attendance'.'/'.$USER->id.'/'.'temp.csv';
+        $header = ['ID Number', 'Symbology', 'Scan Time', 'Scan Date', 'Current Time', 'Current Date', 'Serial Number'];
```
you should look at how the other moodle csv import functions work (like uploadcourse) and make this consistent with the other core csv import tools - it uses strtotime() to convert the content in the passed date field.

jonathanchan2 replied:
I assume you are specifically referring to the string array in line 146. This list / string array is intended to go at the top of the scan file that is generated by the OPN-200X barcode scanner. 

Interesting thing about this scanner and its software is that when you pull the barcode data off it, it does not include the headings to each column. Now, the way the csvreader->next() function work is it skips the first row of data, assuming it is a header row, and starts reading csv data from the second row. This means the first row of attendance data that was pulled from the barcode scanner would be skipped. We can't have that. So to fix this issue, I wrote lines 144 to 173 to insert a header row (the string array you see in line 146) at the top of the uploaded attendance file. This code is necessary for the operation of this feature. 

The string array in line 146 is simply that, a string. Hence, why I capitalized each word of each heading that was to be inserted at the top of the uploaded file. The ID Number column in the scanner file contains the university-recognized student ID number that is encoded by a barcode on each student's student ID card. This is not the student's user ID number on moodle. The student's user ID number is different from his/her student ID number. It is assumed that each university would store each student's unique university-recognized student ID under the "ID number" field when creating a student's moodle profile. This is number used by this feature to determine their attendance.

danmarsden replied:
the problem is that this becomes very specific to your use-case - for this to be accepted into the core atendance module it needs to be more generic and needs to use normal Moodle identifiers.

I'd suggest you take a look at the upload user and  some of the auth/enrol plugins and how they match users against external data - quite often the admin will be able to choose from username/email/idnumber rather than user idnumber which is an optional field and not all sites will use. we also need the "language" to be the same as moodles normal fields. Make sense?

jonathanchan2 replied:
I agree that there is an issue with the specificity of my solution with respect to the fact that I mapped the "ID Number" column in the csv file to the "ID number" field in moodle, which is a very specific optional field that many sites may not use. That is what I am getting from your above response. 

My proposed solution is to create a page that will appear after the user uploads a csv file. This new page will show a preview of the uploaded file and allow the user to define which column of the csv file maps to which field in moodle, sort of how the import grades feature operates. Is this more along the lines of what you're looking for? 

![image](https://user-images.githubusercontent.com/45237038/57867759-b3ccf480-77cf-11e9-82b4-0cdbb6ce8274.png)

Also, could you be more specific when you say "we need the language to be the same as moodle normal fields"?

danmarsden replied:
Thanks - that sounds much better.

I mean that the strings related to user fields need to match the same language in core - eg there is nothing in mooodle called "Symbology" and on the user profile page the user id number field is referred to as "ID number" - not "ID Number" - does that make more sense? 

thanks!

danmarsden replied:
Although - I'm not quite convinced we need "2" drop downs for the user identity and would be happy with just looking at the column names passed in the first row (idnumber/username/email) and then dealing with that automatically but I haven't looked recently at the UI for the import user tool in core moodle.

jonathanchan2 replied:
The purpose of the string $header is to act as a header row since the OPN2001 scanner does not generate its own header row, ie the first row of the file contains attendance data. This was needed because the csv import library function next() assumes the first row of any uploaded csv file is a header row and hence we would lose the first row of attendance data it this was not compensated for before processing the file. $header serves no other purpose than to act as a header row. 

I am thinking of removing the strings in the $header variable and leaving it empty to act as a blank row at the top of the file. This is because other persons may use a different format for the csv file, eg the scantime column might be the first column instead of id number. In this case we would need the 2 drop downs so that the user can identify which column in the file contains the user's identity and then link that to the relevant moodle field for processing.

danmarsden replied:
I'd prefer to keep the header row and automatically select the field based on the name in the header field. This also means that in future we could add an automated cli option rather than relying on a web based upload- but I'm happy with either option in the first release of this.

jonathanchan2 commented on classes/structure.php:
```diff
@@ -85,6 +85,9 @@ class mod_attendance_structure {
     /** @var int Position for the session detail columns related to summary columns.*/
     public $sessiondetailspos;
 
+    /** @var boolean flag set when user attempts to change student's attendance from P,L or E to A */
```
I seriously forgot about removing this. You will see that it does nothing as of right now. But I will proceed to remove instances of it. Thank you for pointing this out.

jonathanchan2 commented on classes/structure.php:
```diff
@@ -712,7 +758,8 @@ public function take_from_form_data($formdata) {
                         $dbsesslog[$log->studentid]->statusid <> $log->statusid ||
                         $dbsesslog[$log->studentid]->statusset <> $log->statusset) {
-
+                        // To prevent users from changing the attendance of studentswho are marked as P,L or E,
```
In fact, only the comment needs to be removed. The code that was supposed to work as the comment say was reverted to what you see below.

jonathanchan2 commented on lang/en/attendance.php:
```diff
@@ -41,6 +41,7 @@
 $string['allcourses'] = 'All courses';
 $string['allpast'] = 'All past';
 $string['allsessions'] = 'All sessions';
+$string['attemptedfraud'] = 'Cannot change the attendance of students who are already marked as Present, Late or Excused.';
```
This was another instance of something that needed to be removed which I forgot about. It will be removed.

jonathanchan2 commented on classes/structure.php:
```diff
@@ -1290,4 +1348,24 @@ public function get_lowgrade_threshold() {
 
         return $this->lowgradethreshold;
     }
+
```
I understand your point. It is not needed and will therefore be changed.

jonathanchan2 commented on add_form.php:
```diff
@@ -280,6 +280,14 @@ public function definition() {
         }
 
+        // Upload attendance files.
+        $mform->addElement('header', 'headeruploadattendance', get_string('uploadattendance', 'attendance'));
```
At the university I work at, lecturers allow their students to scan their student ids (take their own attendance) up to 30 mins before the start of class. With the current plugin, that is not possible for students to perform. Students could only take their attendance when the session has started. 

This setting is meant to cater to these early students by allowing the lecturer to set how early the earliest permissible scan can occur. I chose to make it a setting at the session creating level because here the lecturer only has to set this "earliest time" once for the course if he/she decides to change it from the default 30 mins. I believe this is a necessary option, since many of our lecturers have expressed that they would like the attendance taking process to occur as quickly as possible before class starts so that it does not become a distraction to students.

danmarsden replied:
Right - so that relates to issue #297 and has nothing to do with the import feature so should be submitted as a separate pull request/improvement. It's also not implemented quite right so can you please remove it from this patch and we can then deal with it later as a separate improvement? thanks!

jonathanchan2 replied:
I understand. It will be removed.

jonathanchan2 commented on lang/en/attendance.php:
```diff
@@ -240,14 +247,20 @@
 $string['indetail'] = 'In detail...';
 $string['invalidaction'] = 'You must select an action';
+$string['invalidcontentformat'] = 'The contents of the uploaded csv file are not organised in the format:
+<br>ID Number, Symbology, Scan Time (h:m:s AM/PM), Scan Date (m/d/y), Current Time (h:m:s AM/PM), Current Date (m/d/y), Serial Number
```
I mentioned earlier the reason why these string headings were capitalized the way they were. I understand it is not conventional with moodle's naming convention for fields. However, this is a string warning that will be displayed to the user, indicating that he/she should ensure that the uploaded file is formated in the previously mentioned format of ID number, Symbology, .....

jonathanchan2 replied:
I looked through my code an noticed this string was not being used anywhere else. So I will remove it.

danmarsden commented:
thanks Jonathan, I'm happy to take another look when you've tidied this up a bit more - in particular your import class needs to be a lot more generic and more like other normal moodle import processes rather than the specific format provided by your barcode scanner to make it more useful for other organisations. Thanks!

jonathanchan2 commented:
Hi Dan,

I have completed the major changes including the file preview feature where the user can identify which moodle field they want to identify the student by. Please have an in-depth look at the work I've done and let me know you comments. It would be greatly appreciated.

jonathanchan2 commented:
Hi Dan,

There seems to be an issue with the travis-ci checker. It says the "apt-get install failed". Can you please look into this at your earliest convenience. 

Thank you. 

Best regards,
Jonathan.

jonathanchan2 commented:
I also recall from one of the early travis-ci test failure that there were issues with the behat test. Now, I am not very familiar with this and have not paid much attention to it during my development process. Is it possible for you to provide guidance with this when the travis-ci report comes back for the latest version of my code. 

Thank you, 
Jonathan.

danmarsden commented:
Hi Jonathan - looking at your pull request the "files changed" by this pull request seems to revert a number of changes to the master branch - breaking a number of improvements/fixes. - I'd also try removing your custom changes to the travis.yml file as the core one in the master branch has been updated and your custom version of the travis config might be the reason for the failure here.

Once you have rebased this again can you flick through the diff and make sure it's only changing files that relate to this patch and not adding/changing code unrelated to your upload attendance feature? - Please let me know if you are unable to tidy this up and I can try and fix your patch up manually on my end.

thanks,

danmarsden commented:
- I do wonder if some of this might be tidied up by rebasing your branch against the master branch.

jonathanchan2 commented:
I accidentally deleted the online branch and I am working on trying to get the pull request back up. Do you know how to easily do this?

jonathanchan2 commented:
I decided to open a new pull request and reference this one so that we can continue seeing the upload attendance feature to the end of the development process. 

Thank you for understanding,
Jonathan.
